### ISSUE TITLE
`core.script-args` btest fails sporadically

### ISSUE THREAD

**Issue by timwoj** [zeek/zeek#3774]
I've seen this a few times on CI but I haven't ever been able to reproduce it locally:

```
[#4] core.script-args ... failed on retry #1
  % 'btest-diff out' failed unexpectedly (exit code 1)
  % cat .diag
  == File ===============================
  [-a, -b, -c]
  [-b --, /zeek/testing/btest/.tmp/core.script-args/test.zeek, -d, -e, -f]
  == Diff ===============================
  --- /dev/fd/63	2024-06-07 19:39:25.854387708 +0000
  +++ /dev/fd/62	2024-06-07 19:39:25.853387710 +0000
  @@ -1,2 +1,2 @@
   [-a, -b, -c]
  -[-d, -e, -f]
  +[-b --, /zeek/testing/btest/.tmp/core.script-args/test.zeek, -d, -e, -f]
  =======================================
  % cat .stderr
```

**Comment by ckreibich** [comment]
For context, [here](https://cirrus-ci.com/task/4524291448897536?logs=test#L1978) is a failing one, in a tsan build. There's really nothing racy in the script args logic, and it's pretty baffling how the codepath would lead to the failing output. I've tried a few things and haven't been able to trigger it.

If this keeps happening we should keep an eye on whether it's always in tsan builds. Maybe it somehow affects shebang handling in a bad way.

**Comment by awelzel** [comment]
Just to keep it alive - failed just today again: https://cirrus-ci.com/task/6208855330258944?logs=test#L1968
```
[#3] core.script-args ... failed on retry #1
  % 'btest-diff out' failed unexpectedly (exit code 1)
  % cat .diag
  == File ===============================
  [-a, -b, -c]
  [-b --, /zeek/testing/btest/.tmp/core.script-args/test.zeek, -d, -e, -f]
  == Diff ===============================
  --- /dev/fd/63	2024-08-06 18:42:30.091297018 +0000
  +++ /dev/fd/62	2024-08-06 18:42:30.090297019 +0000
  @@ -1,2 +1,2 @@
   [-a, -b, -c]
  -[-d, -e, -f]
  +[-b --, /zeek/testing/btest/.tmp/core.script-args/test.zeek, -d, -e, -f]
  =======================================
  % cat .stderr
[#5] supervisor.revive-stem ... ok
[#3] core.script-args ... failed on retry #2
  % 'btest-diff out' failed unexpectedly (exit code 1)
  % cat .diag
  == File ===============================
  [-a, -b, -c]
  [-b --, /zeek/testing/btest/.tmp/core.script-args/test.zeek, -d, -e, -f]
  == Diff ===============================
  --- /dev/fd/63	2024-08-06 18:42:31.389295858 +0000
  +++ /dev/fd/62	2024-08-06 18:42:31.388295859 +0000
  @@ -1,2 +1,2 @@
   [-a, -b, -c]
  -[-d, -e, -f]
  +[-b --, /zeek/testing/btest/.tmp/core.script-args/test.zeek, -d, -e, -f]
  =======================================
  % cat .stderr
```

**Comment by JustinAzoff** [comment]
the weird thing about this is that it seems super hard to reproduce, but when it fails it fails 3 times in a row?

Also weird that based on how the options code works, if it puts the wrong thing in `script_args` it should also be putting the wrong thing in `scripts_to_load` which would make the test not even run at all.

Only thing I've noticed that is a bit odd is how the test runs

```
# @TEST-EXEC: $(dirname %INPUT)/test.zeek -d -e -f >>out
```

when it could just be this for consistency

```
# @TEST-EXEC: ./test.zeek -d -e -f >>out
```

Since the test.zeek is written using

```
printf '#!' > test.zeek
```

and not

```
printf '#!' > $(dirname %INPUT)/test.zeek
```

**Comment by awelzel** [comment]
I didn't reproduce locally for me, but it does fail very often on Cirrus when logging into the shell.

TLDR: On Cirrus 

The curious part that when printing `argc[i]` already in main, it contains `-b ./test.zeek` *twice* for a failing run.
```
root@cirrus-ci-task-6130857516204032:/zeek/testing/btest/.tmp/core.script-args# ./test.zeek -d -e -f
main: argv[0]=/zeek/testing/btest/../../build/src/zeek
main: argv[1]=-b --
main: argv[2]=./test.zeek
main: argv[3]=-b --
main: argv[4]=./test.zeek
main: argv[5]=-d
main: argv[6]=-e
main: argv[7]=-f
no test 8
parse_cmdline: argv[0]=/zeek/testing/btest/../../build/src/zeek
parse_cmdline: argv[1]=-b --
parse_cmdline: argv[2]=./test.zeek
parse_cmdline: argv[3]=-b --
parse_cmdline: argv[4]=./test.zeek
parse_cmdline: argv[5]=-d
parse_cmdline: argv[6]=-e
parse_cmdline: argv[7]=-f
[-b --, ./test.zeek, -d, -e, -f]
```

This is a good run:
```
root@cirrus-ci-task-6130857516204032:/zeek/testing/btest/.tmp/core.script-args# ./test.zeek -d -e -f
main: argv[0]=/zeek/testing/btest/../../build/src/zeek
main: argv[1]=-b --
main: argv[2]=./test.zeek
main: argv[3]=-d
main: argv[4]=-e
main: argv[5]=-f
no test 6
parse_cmdline: argv[0]=/zeek/testing/btest/../../build/src/zeek
parse_cmdline: argv[1]=-b --
parse_cmdline: argv[2]=./test.zeek
parse_cmdline: argv[3]=-d
parse_cmdline: argv[4]=-e
parse_cmdline: argv[5]=-f
[-d, -e, -f]
```

The `-b --` and `script name` appear twice in `argv[]` in the failing run, but just once in a successful run.

Adding an exit_only_after_terminate=T and, `ps` also has the duplicated value.
```
root@cirrus-ci-task-6130857516204032:/zeek/testing/btest# ps -ax  -opid,comm,cmd | grep zeek
 323406 test.zeek       /zeek/testing/btest/../../build/src/zeek -b -- ./test.zeek -b -- ./test.zeek -a -b -c
```


Uhm, running with `strace`, the failing run has one extra `execve()` which isn't there when the test passes (and that one looks suspicious):
```
read(3, "", 13010)                      = 0
close(3)                                = 0
munmap(0x78f286c7e000, 32768)           = 0
personality(PER_LINUX|ADDR_NO_RANDOMIZE) = 0 (PER_LINUX)
execve("./test.zeek", ["/zeek/testing/btest/../../build/"..., "-b --", "./test.zeek", "-d", "-e", "-f"], 0x7ffccc806c00 /* 96 vars */) = 0
brk(NULL)       
```

Hmm, and that seems to come from TSAN code, it attempts to turn ASLR off if the memory layout isn't fitting.

https://github.com/llvm/llvm-project/blob/15dacb452f3ae210837f5fc0839489d9801c8fc3/compiler-rt/lib/tsan/rtl/tsan_platform_linux.cpp#L218

IIUC, threads sanitizer re-execs if the memory layout isn't fitting
```
  if (reexec) {
    // Don't check the address space since we're going to re-exec anyway.
  } else if (!CheckAndProtect(false, ignore_heap, false)) {
    // ASLR personality check.
    // N.B. 'personality' is sometimes forbidden by sandboxes, so we only call
    // this as a last resort (when the memory mapping is incompatible and TSan
    // would fail anyway).
    int old_personality = personality(0xffffffff);
    bool aslr_on =
        (old_personality != -1) && ((old_personality & ADDR_NO_RANDOMIZE) == 0);

    if (aslr_on) {
      // Disable ASLR if the memory layout was incompatible.
      // Alternatively, we could just keep re-execing until we get lucky
      // with a compatible randomized layout, but the risk is that if it's
      // not an ASLR-related issue, we will be stuck in an infinite loop of
      // re-execing (unless we change ReExec to pass a parameter of the
      // number of retries allowed.)
      VReport(1,
              "WARNING: ThreadSanitizer: memory layout is incompatible, "
              "possibly due to high-entropy ASLR.\n"
              "Re-execing with fixed virtual address space.\n"
              "N.B. reducing ASLR entropy is preferable.\n");
      CHECK_NE(personality(old_personality | ADDR_NO_RANDOMIZE), -1);
      reexec = true;
```

And the reexec executes re-executes the script, including the shebang line and that's how we end up with the duplicated `-b --`  `test.zeek` part.

With that, I think we just disable the test when running under TSAN? I'm not even sure that's worth a ticket for TSAN.

**Comment by JustinAzoff** [comment]
ah! and related to running in a sandbox, which makes sense. I was starting to think it must be environmental, why else would it never fail for me (I ran it thousands of times) and fail 3x in a row in CI.

I could see tsan not caring about zeek, but wouldn't this also affect python or any other scripting language?

**Comment by awelzel** [comment]
>  I was starting to think it must be environmental, 

Can provoke it with this before (though using a different path): `ulimit -s unlimited`

> I could see tsan not caring about zeek, but wouldn't this also affect python or any other scripting language?

I guess yes? Though only if testing is done via a `.py` shebanged script, rather than `python x.py` which I thought might be unusual.

---

@thurstond / @vitalybuka - we've run into testing failures due TSAN's reexec logic when a shebang script is involved - would you be interested in a ticket for LLVM?

Essentially the reexec cause unexpected duplicate entries in `argv[]` to appear when a shebanged script is involved - here's a short reproducer:
```
$ cat main.c
#include <stdio.h>

int main(int argc, const char *argv[]) {
  printf("argc=%d\n", argc);
  for ( int i = 0; i < argc; i++) {
    printf("argv[%d]=%s\n", i, argv[i]);
  }
}

$ cat main.sh 
#!./main -b --

$ clang-18 -fsanitize=thread main.c -o main 

$ ./main.sh -a
argc=4
argv[0]=./main
argv[1]=-b --
argv[2]=./main.sh
argv[3]=-a

# force a tsan re-exec
$ ulimit -s unlimited

$ ./main.sh -a
argc=6
argv[0]=./main
argv[1]=-b --           # duplicate
argv[2]=./main.sh       # duplicate
argv[3]=-b --
argv[4]=./main.sh
argv[5]=-a
```
